hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenFreezeAccountHandler.java

```diff
@@ -44,23 +54,99 @@ public TokenFreezeAccountHandler() {
-     * change.
-     *
+     * @param txn the {@link TokenFreezeAccountTransactionBody} of the active transaction
+     * @param accountStore the {@link ReadableAccountStore} for the active transaction
+     * @param tokenStore the {@link ReadableTokenStore} for the active transaction
+     * @param tokenRelStore the {@link WritableTokenRelationStore} for the active transaction
      * @throws NullPointerException if one of the arguments is {@code null}
      */
-    public void handle() {
-        throw new UnsupportedOperationException("Not implemented");
+    public void handle(
+            @NonNull final TransactionBody txn,
+            @NonNull ReadableAccountStore accountStore,
+            @NonNull ReadableTokenStore tokenStore,
+            @NonNull WritableTokenRelationStore tokenRelStore)
+            throws HandleException {
+        requireNonNull(txn);
+        requireNonNull(accountStore);
+        requireNonNull(tokenStore);
+        requireNonNull(tokenRelStore);
+
+        final var op = txn.tokenFreezeOrThrow();
+        final var tokenRel = validateSemantics(op, accountStore, tokenStore, tokenRelStore);
+
+        final var copyBuilder = tokenRel.copyBuilder();
+        copyBuilder.frozen(true);
+        tokenRelStore.put(copyBuilder.build());
+    }
+
+    /**
+     * Performs checks independent of state or context
+     */
+    private void pureChecks(TokenFreezeAccountTransactionBody op) throws PreCheckException {
+        if (!op.hasToken()) {
+            throw new PreCheckException(INVALID_TOKEN_ID);
+        }
+
+        if (!op.hasAccount()) {
+            throw new PreCheckException(INVALID_ACCOUNT_ID);
+        }
+    }
+
+    /**
+     * Performs checks that the given token and accounts from the state are valid and that the
+     * token is associated to the account
+     *
+     * @return the token relation for the given token and account
+     */
+    private TokenRelation validateSemantics(
+            TokenFreezeAccountTransactionBody op,
+            ReadableAccountStore accountStore,
+            ReadableTokenStore tokenStore,
+            WritableTokenRelationStore tokenRelStore)
+            throws HandleException {
+        // Check that the token exists
+        final var tokenId = op.tokenOrElse(TokenID.DEFAULT);
+        final ReadableTokenStore.TokenMetadata token;
+        try {
+            token = tokenStore.getTokenMeta(tokenId);
+        } catch (PreCheckException e) {
+            throw new HandleException(INVALID_TOKEN_ID);
+        }
+        if (token == null) {
+            throw new HandleException(INVALID_TOKEN_ID);
+        }
+        if (!token.hasFreezeKey()) {
+            throw new HandleException(TOKEN_HAS_NO_FREEZE_KEY);
+        }
+
+        // Check that the account exists
+        final var account = accountStore.getAccountById(op.accountOrElse(AccountID.DEFAULT));
+        if (account == null) {
+            throw new HandleException(INVALID_ACCOUNT_ID);
```

Comment:
we could use `validateTrue(account != null, INVALID_ACCOUNT_ID)`

Reply:
Done